crates/transport/src/layers/retry.rs

```diff
@@ -0,0 +1,276 @@
+use crate::{
+    error::{TransportError, TransportErrorKind},
+    TransportFut,
+};
+use alloy_json_rpc::{ErrorPayload, RequestPacket, ResponsePacket};
+use serde::Deserialize;
+use std::{
+    sync::{
+        atomic::{AtomicU32, Ordering},
+        Arc,
+    },
+    task::{Context, Poll},
+};
+use tower::{Layer, Service};
+use tracing::trace;
+
+/// A Transport Layer that is responsible for retrying requests based on the
+/// error type. See [`TransportError`].
+///
+/// TransportError: crate::error::TransportError
+#[derive(Debug, Clone)]
+pub struct RetryBackoffLayer {
+    /// The maximum number of retries for rate limit errors
+    max_rate_limit_retries: u32,
+    /// The maximum number of retries for timeout errors
+    max_timeout_retries: u32,
+    /// The initial backoff in milliseconds
+    initial_backoff: u64,
+    /// The number of compute units per second for this provider
+    compute_units_per_second: u64,
+}
+
+impl RetryBackoffLayer {
+    /// Creates a new retry layer with the given parameters.
+    pub const fn new(
+        max_rate_limit_retries: u32,
+        max_timeout_retries: u32,
+        initial_backoff: u64,
+        compute_units_per_second: u64,
+    ) -> Self {
+        Self {
+            max_rate_limit_retries,
+            max_timeout_retries,
+            initial_backoff,
+            compute_units_per_second,
+        }
+    }
+}
+
+/// [RateLimitRetryPolicy] implements [RetryPolicy] to determine whether to retry depending on the
+/// err.
+#[derive(Debug, Clone)]
+pub struct RateLimitRetryPolicy;
+
+/// [RetryPolicy] defines logic for which [TransportError] instances should
+/// the client retry the request and try to recover from.
+pub trait RetryPolicy: Send + Sync + std::fmt::Debug {
+    /// Whether to retry the request based on the given `error`
+    fn should_retry(&self, error: &TransportError) -> bool;
+
+    /// Providers may include the `backoff` in the error response directly
+    fn backoff_hint(&self, error: &TransportError) -> Option<std::time::Duration>;
+}
+
+impl RetryPolicy for RateLimitRetryPolicy {
+    fn should_retry(&self, error: &TransportError) -> bool {
```

Comment:
this method takes self, but doesn't use it. Instead, this entire method could be implemented on `TransportError`

Reply:
if we wanted to preserve the option for other implementers to change the retry policy, we could move all THIS logic to the `TransportError` and still keep this method here as a simple 

```
    fn should_retry(&self, error: &TransportError) -> bool {
        error.is_retryable()
    }
```